crates/typst/src/layout/align.rs

```diff
@@ -445,6 +476,150 @@ cast! {
     }
 }
 
+/// An internal representation that combines horizontal or vertical alignments. The
+/// allowed alignment positions are designated by the type parameter `H` and `V`.
+///
+/// This is not user-visible, but an internal type to impose type safety. For example,
+/// `SpecificAlignment<HAlignment, OuterVAlignment>` does not allow vertical alignment
+/// position "center", because `V = OuterVAlignment` doesn't have it.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    H(H),
+    V(V),
+    Both(H, V),
+}
+
+impl<H, V> SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
```

Comment:
As a consequence of my previous comment:

```suggestion
impl<H, V> SpecificAlignment<H, V> {
```